rpcs3/rpcs3qt/main_window.cpp

```diff
@@ -1578,46 +1578,73 @@ void main_window::DecryptSPRXLibraries()
+			"\nAnd is logged with some sceNpDrm* functions when the game/application which owns \"%0\" is running.").arg(qstr(filename));
 
-		if (tried)
+		if (repeat_count >= 2)
 		{
-			gui_log.error("Failed to decrypt %s with specfied KLIC, retrying.\n%s", old_path, sstr(hint));
+			gui_log.error("Failed to decrypt %s with specfied KLIC, retrying.\n%s", path, sstr(hint));
 		}
 
-		input_dialog dlg(32, "", tr("Enter KLIC of %0").arg(qstr(path)),
-			tried ? tr("Decryption failed with provided KLIC.\n%0").arg(hint) : tr("Hexadecimal only."), "00000000000000000000000000000000", this);
+		input_dialog* dlg = new input_dialog(39, "", tr("Enter KLIC of %0").arg(qstr(filename)),
+			repeat_count >= 2 ? tr("Decryption failed with provided KLIC.\n%0").arg(hint) : tr("Hexadecimal value."), "KLIC=0x00000000000000000000000000000000", this);
 
 		QFont mono = QFontDatabase::systemFont(QFontDatabase::FixedFont);
 		mono.setPointSize(8);
-		dlg.set_input_font(mono, true, '0');
-		dlg.set_clear_button_enabled(false);
-		dlg.set_button_enabled(QDialogButtonBox::StandardButton::Ok, false);
-		dlg.set_validator(new QRegularExpressionValidator(QRegularExpression("^[a-fA-F0-9]*$"))); // HEX only
+		dlg->set_input_font(mono, true, '0');
+		dlg->set_clear_button_enabled(false);
+		dlg->set_button_enabled(QDialogButtonBox::StandardButton::Ok, false);
+		dlg->set_validator(new QRegularExpressionValidator(QRegularExpression("^(K(L(I(C(=(0x?)?)?)?)?)?)?[a-fA-F0-9]{0,32}$"))); // HEX only (with additional klic=0x prefix for convenience))
+		dlg->setAttribute(Qt::WA_DeleteOnClose);
 
-		connect(&dlg, &input_dialog::text_changed, &dlg, [&dlg](const QString& text)
+		connect(dlg, &input_dialog::text_changed, dlg, [dlg](const QString& text)
 		{
-			dlg.set_button_enabled(QDialogButtonBox::StandardButton::Ok, text.size() == 32);
+			dlg->set_button_enabled(QDialogButtonBox::StandardButton::Ok, text.size() - (text.indexOf('x') + 1) == 32);
 		});
 
-		if (dlg.exec() == QDialog::Accepted)
+		dlg->show();
+
+		connect(dlg, &QDialog::accepted, this, [this, iterate, dlg, vec_modules, mod_index, klics_storage, repeat_count]()
 		{
-			return sstr(dlg.get_input_text());
-		}
+			std::string text = sstr(dlg->get_input_text());
+
+			if (usz new_index = decrypt_sprx_libraries(*vec_modules, mod_index, text, *klics_storage); new_index < vec_modules->size())
+			{
+				QTimer::singleShot(0, [dlg, iterate, vec_modules, mod_index, klics_storage, repeat_count, new_index]()
+				{
+					// Increase repeat count if "stuck" on the same file
+					(*iterate)(new_index, new_index == mod_index ? repeat_count + 1 : 0);
+				});
+			}
+		});
 
-		return {};
+		connect(dlg, &QDialog::rejected, this, []()
+		{
+			gui_log.notice("User has cancelled entering KLIC.");
+		});
 	};
 
-	decrypt_sprx_libraries(vec_modules, input_cb);
+	if (usz new_index = decrypt_sprx_libraries(*vec_modules, 0, "", *klics_storage); new_index < vec_modules->size())
+	{
+		QTimer::singleShot(0, [=]()
```

Comment:
why do we need a singleshot here if the dialog is async now?

Reply:
Overlooked it.